### ISSUE TITLE
check_not_running is not reliable

### ISSUE THREAD

**Issue by Lu5ck** [quarantin/pz-server-tools#31]
Sometimes when I do "pzserver restart", the server shuts down but then it doesn't boot up because it says "Server is running" when it isn't. I had to reboot the vps to get working again.

**Comment by quarantin** [comment]
I know sometimes the pz server will freeze on exit, so it never really quits. Unfortunately this is a bug from pz itself and I can't do much about it. You could still use `pzserver kill` in that case to kill the process.

**Comment by Lu5ck** [comment]
No, the process completely ended, I double checked with top but check_not_running which read PROC will return otherwise.

**Comment by Lu5ck** [comment]
`pzserver kill` doesn't remove it from `/proc` which your script checking at
i think removing from `/proc` would require `sudo` access, right?
if you gonna check `/proc/pid`, perhaps also check `/proc/pid/stat`?
or perhaps use `pgrep`?

PS: I don't know the stat of the /proc when this happened, will look at it when it happen again.

**Comment by Lu5ck** [comment]
According to your debug script, pid doesn't exist so something else is stopping the script from running, I assume.
/var/cache/pzst/pzserver.pzserver1.lock still exist, I assume because of this?

[troubleshoot.log](https://github.com/quarantin/pz-server-tools/files/11581430/troubleshoot.log)

**Comment by quarantin** [comment]
In order to better understand the problem and find ways to reproduce the bug, can you confirm the following:
- A restart was initiated through pzst (either with `pzserver restart`, or with daily restart)
- Someone tried to make the server quit by typing the `/quit` command in the in-game console while a restart was already initiated.
- Someone tried to cancel the restart with `pzserver cancelrestart` (or maybe not, but I really need to know)

A few things to note:

1) Your pzst are not up-to-date. I recently fixed a bug (https://github.com/quarantin/pz-server-tools/commit/faf53155f51e2c35c65591d70b79e3775fc5b465) which I suspect might have an impact on your issue.
Please update pzst.

2) You are using Ubuntu 22.04 which is not supported.
I only support Ubuntu 20.04 for now, as stated here: https://github.com/quarantin/pz-server-tools#supported-plateform.
Please don't open new issues for unsupported OS/version.

3) If for some reasons you really need me to support Ubuntu 22.04, I can do it but it would basically double my workload to maintain pzst because twice as much testing would be needed.
In that case please consider making substantial monthly donations on my Ko-fi to support my efforts:
https://ko-fi.com/quarantin

**Comment by Lu5ck** [comment]
> A restart was initiated through pzst (either with pzserver restart, or with daily restart)
Yes

> Someone tried to make the server quit by typing the /quit command in the in-game console while a restart was already initiated.
No, is a auto restart

> Someone tried to cancel the restart with pzserver cancelrestart (or maybe not, but I really need to know)
No, is auto restart

> Your pzst are not up-to-date. I recently fixed a bug (https://github.com/quarantin/pz-server-tools/commit/faf53155f51e2c35c65591d70b79e3775fc5b465) which I suspect might have an impact on your issue.
Please update pzst.
Just updated to run pzdebug, will see if the problem persist

> You are using Ubuntu 22.04 which is not supported.
> I only support Ubuntu 20.04 for now, as stated here: https://github.com/quarantin/pz-server-tools#supported-plateform.
> Please don't open new issues for unsupported OS/version.
TBH, I could edit the log and you won't know I am using 22.04 ¯\_(ツ)_/¯
Don't worry mate, if this problem is resolved, you can put 22.04 as supported as this literally the only problem! Although you also need to add additional step for installation process as 22.04 does have stricter home directory permission, would need to manually add "pzst" user group to "pzserver1" home directory.

**Comment by quarantin** [comment]
> TBH, I could edit the log and you won't know I am using 22.04 ¯_(ツ)_/¯
Well, I'm not sure if you realize how stupid and immature that answer is.
- First it would do no one any good because I would be spending endless time trying to reproduce a bug in an environment where it probably can't be reproduced, and I would never be able to fix the bug anyway. This would really be cruel and pointless.
- Then, assuming I find out about it somehow (and yes I would have found out about it, there's plenty of subtle ways to fingerprint a specific OS/version), you'd be blacklisted for any kind of support from me.
-  Also I'd probably stop working on this project because I would feel betrayed by my own users, to which I distribute my tools and give support for free.
- Finally, not that I would fall that low, but it's more of a general advice, I'd recommend against pissing off people whose code you're running on your server for obvious security reasons.

**Comment by buffyuwu** [comment]
> TBH, I could edit the log and you won't know I am using 22.04 ¯_(ツ)_/¯ 

![image_2023-05-27_183131961](https://github.com/quarantin/pz-server-tools/assets/42501819/4e2385ee-d2b7-4992-914d-4b01450c9432)

**Comment by Lu5ck** [comment]
> > TBH, I could edit the log and you won't know I am using 22.04 ¯_(ツ)_/¯
> 
> Well, I'm not sure if you realize how stupid and immature that answer is.
> 
> * First it would do no one any good because I would be spending endless time trying to reproduce a bug in an environment where it probably can't be reproduced, and I would never be able to fix the bug anyway. This would really be cruel and pointless.
> * Then, assuming I find out about it somehow (and yes I would have found out about it, there's plenty of subtle ways to fingerprint a specific OS/version), you'd be blacklisted for any kind of support from me.
> * Also I'd probably stop working on this project because I would feel betrayed by my own users, to which I distribute my tools and give support for free.
> * Finally, not that I would fall that low, but it's more of a general advice, I'd recommend against pissing off people whose code you're running on your server for obvious security reasons.

I am just jesting mate. I do code myself and I do know that sometimes you need to reproduce the problem to get a concrete understanding of it, to debug, inorder to fix it.

On a serious note of why

- I assume you are running your linux off either WSL, Hyper-V or dual boot, all of which are local installation based on image taken off official Ubuntu. However, I dare say that much of your users base are running on image based off what their server provider provided which is modified for their server environment security and hardware needs. That by itself is already a environment difference, regardless of OS difference.
- Since you running off a local installation, you pretty have full access to every feature of the linux. However, that is not much of the case for your users base who running off either VPS or rented dedicated server. OS difference is not a critical worries because your script is very dependent on one particular feature which is TIOCSTI. If TIOCSTI is disabled which can happen on some providers, your entire script will stop working properly. I have learnt that the hard way.
- You running on a local installation so I assume you interface with linux differently from much of us. Much of us are interfacing using SSH instead of direct terminal access. This difference will result in how codes returns differently which I had reported it previously and you fixed it. I am just stating this to highlight there is always a difference in our environment regardless of the OS version.
- 22.04 perform better than 20.04. PZ is infamously CPU intensive and every tiny bit of performance improvement matters. This is particularly true for people running on Ryzen. Yes, I am running on Ryzen 5950x and is the 2nd fastest CPU you can find on rental server market. Fastest CPU for server market now is Ryzen 79xx which is still hard to find a provider providing it in my region.
- LTS, 22.04 has 2 longer years support than 20.04, eventually most people gonna switch to it.
- While forward compatibility is not a thing, but backward compatibility should be. 22.04 has stricter security and getting it working on 22.04 will also means 20.04 which has lesser security demand should not have any problem.
- There is only a upgrade path, no downgrade path.
- All in all, I believe you should move on to 22.04 for both performance and security aspect.

**Comment by fmossri** [comment]
Hi, I'm sorry to get in your conversation, but I'm kinda new to github and coding, to I couldn't find how to open my own discussion.
I'm having a similar problem to OP's.
I'm running your app to run my server on a gcloud VM with ubuntu 20.04
I set an automatic reset for this morning, but when I woke up the server was down, and no proccess was running on the VM.
I tried to use pzserver start/restart, but it returns that pzserver is already running. When I try stop/quit/kill it says there's none running. I ran the debug code you posted previously, and it generated a troubleshooting file that I don't know how to read. Do you have any idea what can be happening to my server? :/
